docs/dg/dev/architecture/coding-conventions.md

```diff
@@ -0,0 +1,2165 @@
+- The inherited `getFactory()` method grants access to the [Factory](#factory) layer.
+- The inherited `getFacade()` or `getClient()` methods grant access to the corresponding [facade](#facade-design-pattern) functionalities.
+
+**Example**
+```php
+<?php
+
+namespace Pyz\Zed\ConfigurableBundleGui\Communication\Controller;
+
+/**
+ * @method \Pyz\Zed\ConfigurableBundleGui\Communication\ConfigurableBundleGuiCommunicationFactory getFactory()
+ * @method \Spryker\Zed\ConfigurableBundleGui\Business\ConfigurableBundleGuiFacadeInterface getFacade()
+ */
+class TemplateController extends Spryker\Zed\ConfigurableBundleGui\Communication\Controller\TemplateController
+{
+    /**
+     * @var string
+     */
+    protected const PARAM_ID_CONFIGURABLE_BUNDLE_TEMPLATE = 'id-configurable-bundle-template';
+    
+    /**
+     * @param \Symfony\Component\HttpFoundation\Request $request
+     *
+     * @return \Symfony\Component\HttpFoundation\JsonResponse
+     */
+    public function slotTableAction(Request $request): JsonResponse
+    {
+        $idConfigurableBundleTemplate = $this->castId(
+            $request->get(static::PARAM_ID_CONFIGURABLE_BUNDLE_TEMPLATE),
+        );
+
+        $table = $this->getFactory()->createConfigurableBundleTemplateSlotTable($idConfigurableBundleTemplate);
+
+        return $this->jsonResponse($table->fetchData());
+    }
+}
+```
+
+## Dependency Provider
+
+```
+[Organization]
+├── Zed
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Yves
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Glue
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+├── Client
+│   └── [Module]        
+│       ├── [Module]Config.php
+│       └── [Module]DependencyProvider.php
+└── Service
+    └── [Module]        
+        ├── [Module]Config.php
+        └── [Module]DependencyProvider.php
+```
+
+**Description**
+
+Injects required dependencies to a module application layer. Typically, dependencies are [facades](#facade-design-pattern) or [plugins](#plugin). 
+
+Dependency injection is orchestrated through a `provide-add-get` structure (see example below).
+- The `provide` method is the highest level that holds only `add` calls without any additional logic.
+- The `add` method is the middle level that injects the dependencies into the dependency container using a class constant and a late-binding instantiating closure.
+- The `get` method is the lowest level that sources the dependency.
+
+**Conventions**
+- All class constants MUST be `public` (to decrease conflicts in definition for being a public API class).
+- Each dependency MUST be defined using a class constant.
+- Setting dependency using the `container::set()` MUST be paired with late-binding closure definition to decouple instantiation.
+- Dependencies MUST be wired through the target layer corresponding inherited method:
```

Comment:
Dependencies that required to be instantiated every time we call it, e.g. Queries MUST be added via `Container::factory()` method. E.g. 
`protected function addUserPropelQuery(Container $container): Container
    {
        $container->set(static::PROPEL_QUERY_USER, $container->factory(function (): SpyUserQuery {
            return SpyUserQuery::create();
        }));

        return $container;
    }`